routes/beetles.py

```diff
@@ -0,0 +1,578 @@
+            the name is not given at ingest
+
+    Returns:
+        Nested dict of results for all non-X/Y axis combinations,
+    """
+
+    # operates on underlying DataArray
+    da = ds[bandname]
+
+    # get axis (dimension) names and make list of all coordinate combinations
+    all_dims = da.dims
+    dimnames = [dimname for dimname in all_dims if dimname not in ("X", "Y")]
+    iter_coords = list(
+        itertools.product(*[list(ds[dimname].values) for dimname in dimnames])
+    )
+
+    data_arr = []
+    for coords in iter_coords:
+        sel_di = {dimname: int(coord) for dimname, coord in zip(dimnames, coords)}
+        data_arr.append(da.sel(sel_di).values)
+    data_arr = np.array(data_arr)
+
+    # need to transpose the 2D spatial slices if X is the "rows" dimension
+    if all_dims.index("X") < all_dims.index("Y"):
+        data_arr = data_arr.transpose(0, 2, 1)
+
+    data_arr_mask = np.broadcast_to(poly_mask_arr.mask, data_arr.shape)
+    data_arr[data_arr_mask] = np.nan
+
+    # Adds one to each value to generate correct shape and
+    # for iterations through the data below.
+    eras = sel_di["era"] + 1
+    models = sel_di["model"] + 1
+    scenarios = sel_di["scenario"] + 1
+    snowpacks = sel_di["snowpack"] + 1
+
+    return_arr = np.zeros((eras, models, scenarios, snowpacks))
+
+    # Percentages of each slice of the data. Has a last dimension that are
+    # integer values representing 1 = low risk, 2 = moderate risk, 3 = high risk
+    return_percentages = np.zeros((eras, models, scenarios, snowpacks, 4))
+    for era in range(eras):
+        for model in range(models):
+            for scenario in range(scenarios):
+                for snowpack in range(snowpacks):
+                    index = (
+                        (era * models * scenarios * snowpacks)
+                        + (model * scenarios * snowpacks)
+                        + (scenario * snowpacks)
+                        + snowpack
+                    )
+                    slice = data_arr[index]
+
+                    # Generates a data array that has all NAN values removed
+                    rm_nan_slice = slice[~np.isnan(slice)]
+
+                    # If there is no data in this slice, set the
+                    # value to 0 and continue onto the next data slice.
+                    if len(rm_nan_slice) == 0:
+                        return_arr[era][model][scenario][snowpack] = 0
+                        continue
+
+                    # Generates counts for all 1, 2, or 3 values for beetle risk
+                    uniques = np.unique(rm_nan_slice, return_counts=True)
+
+                    # Sets the "mode" to the first of the unique values
+                    # Must have at least one value to have gotten to this part
+                    # of the script.
+                    mode = uniques[0][0]
+                    mode_count = uniques[1][0]
```

Comment:
You may need to force a sort on `uniques` because the `np.unique` sorts the output, but just in ascending order of the values, not the value counts.

My test:
```python
 x = np.array([1, 2, 2, 3, 3, 3,])

np.unique(x)
Out: array([1, 2, 3])

np.unique(x, return_counts=True)
Out: (array([1, 2, 3]), array([1, 2, 3]))

Reply:
Wait, I see from reading further that you are actually setting a "test mode" in this first part of the code and the logic is:
 - hey, if there is just a single value, then that value is obviously the mode statistic, so set an "initial" or "test" mode to be the first value returned by `np.unique`
 - if there is more than one unique value, get the percentage of pixels that have that value, and sequentially check to see if you have a maximum percentage, and then return the appropriate mode.
 
 I think you can disregard for my first comment, but I'll leave it now for others or just to illustrate how I'm reading this.